File: IT/Containers/Esercizi/Java/Esercizio31.java
/* file Esercizio31.java */

import java.util.Map;
import java.util.Map.Entry;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

public class Esercizio31 {
	public static void main(String [] args) {
		Map<String, String> students = new HashMap<String, String>();

		students.put("A23234", "Mario Rossi");
		students.put("A01357", "Luigi Bianchi");
		students.put("A12579", "Giovanni Verdi");

		Map<String, String> cdl = new HashMap<String, String>();

		cdl.put("A12579", "Informatica");
		cdl.put("A01357", "Matematica");
		cdl.put("A23234", "Fisica");

		System.out.println("Studenti del corso");

		List<Entry<String, String>> arr = new ArrayList<Entry<String, String>>(students.entrySet());

		Collections.sort(arr, new Comparator<Entry<String, String>>() {
				@Override
				public int compare(Entry<String, String> a, Entry<String, String> b) {
					return a.getValue().compareTo(b.getValue());
				}
			}
		);

		for (Entry<String, String> entry : arr) {
			String matr = entry.getKey();
			System.out.println("Matricola " + matr + ", " + entry.getValue() + ", corso di laurea in " + cdl.get(matr) + ".");
		}
	}
}

File: IT/1/Containers/Esercizi/C++/esercizio31.cpp
/*
file esercizio31.cpp
compilare con: g++ -std=c++11 esercizio31.cpp
*/

#include <iostream>
#include <algorithm>
#include <unordered_map>
#include <vector>

using std::cout;
using std::endl;
using std::string;
using std::pair;
using std::unordered_map;
using std::vector;
using std::sort;

bool comp(const pair<string, string>& p1, const pair<string, string>& p2) {
	return p1.second < p2.second;
}

int main()
{
	unordered_map<string, string> students = {{"A23234", "Mario Rossi"}, {"A01357", "Luigi Bianchi"}, {"A12579", "Giovanni Verdi"}};

	unordered_map<string, string> cdl = {{"A01357", "Matematica"}, {"A12579", "Informatica"}, {"A23234", "Fisica"}};

	cout << "Studenti del corso" << endl;

	/* la mappa non permette l'ordinamento, creiamo un vector per applicare l'ordinamento voluto */

	vector<pair<string, string>> vect(students.begin(), students.end());
	sort(vect.begin(), vect.end(), comp);

	for (auto& student : vect) {
		string matr = student.first;
		cout << "Matricola " << matr << ", " << student.second << ", corso di laurea in " << cdl[matr] << "." << endl;
	}
}

